In [24]:
import numpy as np 
import pandas as pd

In [25]:
df = pd.concat(map(pd.read_csv, ['Brazil.csv', 'germany.csv','germany1.csv', 
                                 'Portugal1.csv', 'Spain1.csv', 'Switzerland.csv',
                                'Switzerland1.csv', 'France1.csv',
                                'Brazil1.csv', 'Netherlands.csv', 
                                'Belgium.csv',  'England.csv',
                                'Argentina.csv', 'Uruguay.csv', 
                                'Senegal.csv', 'Ecuador.csv', 'CostaRica.csv',
                                'Mexico.csv', 'France.csv', 'Denmark.csv',
                                'Australia.csv', 'Serbia.csv',
                                'Cameroon.csv', 'Ghana.csv', 'Iran.csv',
                                'Poland.csv', 'SaudI.csv', 'Tunisia.csv', 
                                 'Japan.csv', 'Croatia.csv', 'Morocco.csv',
                                 'Spain2.csv', 'Portugal2.csv', 
                                 'England2.csv', 'Belgium2.csv', 'Uruguay2.csv',
                                'Canada2.csv', 'Netherlands2.csv', 'Saudi2.csv', 
                                'Croatia2.csv', 'Cameroon2.csv', 'Denmark2.csv',
                                'Costa2.csv', 'Morocco2.csv', 'Poland2.csv',
                                'Tunisia2.csv', 'Japan2.csv', 'Argentina2.csv', 
                                'Slovakia.csv', 'Iraq.csv', 'Jamaica.csv',
                                'Kuwait.csv', 'ElSalvador.csv', 'China.csv',
                                'Turkey.csv', 'Ukraine.csv', 'Bolivia.csv',
                                'Togo.csv', 'Togo.csv', 'Angola.csv', 
                                  'Italy.csv', 'Paraguay.csv', 
                                 'Russia.csv',
                                'Algeria.csv', 'Chile.csv', 'Yugoslavia.csv',
                                'Scotland.csv', 'Romania.csv', 'Greece.csv',
                                'Honduras.csv', 'Austria.csv', 'Bulgaria.csv', 
                                 'IvoryCoast.csv', 'SouthAfrica.csv', 'RepIreland.csv',
                                'Slovenia.csv', 'Czechoslovakia.csv', 'Hungary.csv', 
                                'Egypt.csv', 'Norway.csv', 'Peru.csv',
                                'NewZealand.csv', 'Bosnia.csv', 'Panama.csv','UAE.csv',
                                'Iceland.csv', 'USA.csv', 'USA1.csv',  'Nigeria.csv',
                                 'Colombia.csv', 'Sweden.csv']))



In [26]:
# 158 with wc 2022 teams

In [27]:
df.shape

(263, 18)

In [28]:
df.duplicated().value_counts()

False    262
True       1
dtype: int64

In [29]:
df = df.drop_duplicates()

In [30]:
df.shape

(262, 18)

In [31]:
df

Unnamed: 0   goals_z      xg_z  crosses_z  boxtouches_z  passes_z  \
0    Brazil98  0.423077  0.146923  -0.136154     -0.030000  0.429231   
1    Brazil94  0.216923  0.348462   0.031538      0.158462  0.835385   
2    Brazil90  0.113846  0.392308   0.599231      0.513846  0.833077   
3    Brazil86  0.479231  0.609231   0.227692      0.450769  0.770769   
4    Brazil82  0.877692  0.773846   0.428462      0.659231  0.754615   
..        ...       ...       ...        ...           ...       ...   
0    Sweden18  0.110000  0.338462  -0.006923      0.064615 -0.353077   
1    Sweden06  0.013846  0.336923   0.486154      0.567692  0.013077   
2    Sweden02  0.152308  0.160769   0.123077      0.047692 -0.213846   
3    Sweden94  0.437692  0.183077   0.188462      0.307692 -0.013846   
4    Sweden90  0.115385 -0.049231   0.040000      0.268462  0.065385   

    progpasses_z  takeons_z  progruns_z  tackles_z  interceptions_z  \
0       0.037692   0.244615   -0.220000   0.216154         0.270000   
1       0.626923   0.270000    0.266923   1.143846         0.834615   
2       0.405385   0.521538    1.139231   1.109231         0.859231   
3       0.042308   0.337692    0.927692   0.506923         1.015385   
4       0.335385   0.023077    0.638462   0.493846         0.637692   
..           ...        ...         ...        ...              ...   
0      -0.141538  -0.211538   -0.300769  -0.573846        -0.342308   
1       0.272308   0.076923    0.202308   0.139231        -0.020000   
2      -0.001538  -0.110000   -0.378462  -0.200000        -0.078462   
3       0.101538  -0.116154    0.056154  -0.348462        -0.208462   
4       0.035385  -0.134615   -0.349231  -0.089231         0.016923   

    clearances_z  blocks_z  aerials_z   fouls_z  fouled_z    nsxg_z  results  
0      -0.076923 -0.097692  -0.020000 -0.224615  0.100769 -0.124615        2  
1      -0.059231 -0.134615   0.196923 -0.016923  0.033846  0.146923        1  
2       0.084615  0.103077   0.683846  0.213846  0.353846  0.786923        9  
3       0.020769  0.381538   0.038462  0.039231  0.022308  0.692308        5  
4      -0.117692 -0.033846   0.572308 -0.016154 -0.096923  0.890769        5  
..           ...       ...        ...       ...       ...       ...      ...  
0       0.558462  0.088462   0.298462  0.081538  0.203846  0.050769        7  
1      -0.065385 -0.088462   0.296154 -0.152308 -0.132308  0.631538       14  
2       0.894615 -0.050769   0.336923  0.004615  0.038462 -0.066154       13  
3       0.123077  0.126923   0.848462 -0.230000  0.046923  0.230769        3  
4      -0.153077 -0.200769   0.652308 -0.073846  0.069231  0.133846       21  

[262 rows x 18 columns]

# convert to csv

In [32]:
# df = df.drop(columns=['Unnamed: 0'])

In [33]:
df

Unnamed: 0   goals_z      xg_z  crosses_z  boxtouches_z  passes_z  \
0    Brazil98  0.423077  0.146923  -0.136154     -0.030000  0.429231   
1    Brazil94  0.216923  0.348462   0.031538      0.158462  0.835385   
2    Brazil90  0.113846  0.392308   0.599231      0.513846  0.833077   
3    Brazil86  0.479231  0.609231   0.227692      0.450769  0.770769   
4    Brazil82  0.877692  0.773846   0.428462      0.659231  0.754615   
..        ...       ...       ...        ...           ...       ...   
0    Sweden18  0.110000  0.338462  -0.006923      0.064615 -0.353077   
1    Sweden06  0.013846  0.336923   0.486154      0.567692  0.013077   
2    Sweden02  0.152308  0.160769   0.123077      0.047692 -0.213846   
3    Sweden94  0.437692  0.183077   0.188462      0.307692 -0.013846   
4    Sweden90  0.115385 -0.049231   0.040000      0.268462  0.065385   

    progpasses_z  takeons_z  progruns_z  tackles_z  interceptions_z  \
0       0.037692   0.244615   -0.220000   0.216154         0.270000   
1       0.626923   0.270000    0.266923   1.143846         0.834615   
2       0.405385   0.521538    1.139231   1.109231         0.859231   
3       0.042308   0.337692    0.927692   0.506923         1.015385   
4       0.335385   0.023077    0.638462   0.493846         0.637692   
..           ...        ...         ...        ...              ...   
0      -0.141538  -0.211538   -0.300769  -0.573846        -0.342308   
1       0.272308   0.076923    0.202308   0.139231        -0.020000   
2      -0.001538  -0.110000   -0.378462  -0.200000        -0.078462   
3       0.101538  -0.116154    0.056154  -0.348462        -0.208462   
4       0.035385  -0.134615   -0.349231  -0.089231         0.016923   

    clearances_z  blocks_z  aerials_z   fouls_z  fouled_z    nsxg_z  results  
0      -0.076923 -0.097692  -0.020000 -0.224615  0.100769 -0.124615        2  
1      -0.059231 -0.134615   0.196923 -0.016923  0.033846  0.146923        1  
2       0.084615  0.103077   0.683846  0.213846  0.353846  0.786923        9  
3       0.020769  0.381538   0.038462  0.039231  0.022308  0.692308        5  
4      -0.117692 -0.033846   0.572308 -0.016154 -0.096923  0.890769        5  
..           ...       ...        ...       ...       ...       ...      ...  
0       0.558462  0.088462   0.298462  0.081538  0.203846  0.050769        7  
1      -0.065385 -0.088462   0.296154 -0.152308 -0.132308  0.631538       14  
2       0.894615 -0.050769   0.336923  0.004615  0.038462 -0.066154       13  
3       0.123077  0.126923   0.848462 -0.230000  0.046923  0.230769        3  
4      -0.153077 -0.200769   0.652308 -0.073846  0.069231  0.133846       21  

[262 rows x 18 columns]

In [34]:
# df.to_csv('./WC Data Models/WC.csv')


In [121]:
from sklearn.linear_model import LinearRegression
from sklearn.inspection import permutation_importance
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge



In [70]:
df.head()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  progpasses_z  \
0  0.423077  0.146923  -0.136154     -0.030000  0.429231      0.037692   
1  0.216923  0.348462   0.031538      0.158462  0.835385      0.626923   
2  0.113846  0.392308   0.599231      0.513846  0.833077      0.405385   
3  0.479231  0.609231   0.227692      0.450769  0.770769      0.042308   
4  0.877692  0.773846   0.428462      0.659231  0.754615      0.335385   

   takeons_z  progruns_z  tackles_z  interceptions_z  clearances_z  blocks_z  \
0   0.244615   -0.220000   0.216154         0.270000     -0.076923 -0.097692   
1   0.270000    0.266923   1.143846         0.834615     -0.059231 -0.134615   
2   0.521538    1.139231   1.109231         0.859231      0.084615  0.103077   
3   0.337692    0.927692   0.506923         1.015385      0.020769  0.381538   
4   0.023077    0.638462   0.493846         0.637692     -0.117692 -0.033846   

   aerials_z   fouls_z  fouled_z    nsxg_z  results  
0  -0.020000 -0.224615  0.100769 -0.124615        2  
1   0.196923 -0.016923  0.033846  0.146923        1  
2   0.683846  0.213846  0.353846  0.786923        9  
3   0.038462  0.039231  0.022308  0.692308        5  
4   0.572308 -0.016154 -0.096923  0.890769        5

In [16]:
y = df['results']

In [17]:
X = df.drop(columns=['results'])

In [18]:
X_train, X_test, y_train, y_test = train_test_split(
     X, y, test_size=0.20, random_state=42)

In [19]:
model = Ridge(alpha=1e-2).fit(X_train, y_train)
model.score(X, y)

0.4012771300488499

In [20]:
r = permutation_importance(model, X, y, n_repeats=30, random_state=0)

# Permutation feature importance

In [21]:
scoring = ['r2', 'neg_mean_absolute_percentage_error', 'neg_mean_squared_error']
r_multi = permutation_importance(
     model, X, y, n_repeats=30, random_state=0, scoring=scoring)

for metric in r_multi:
     print(f"{metric}") 
     r = r_multi[metric]
     for i in r.importances_mean.argsort()[::-1]:
            if r.importances_mean[i] - 2 * r.importances_std[i] > 0:
                print(f"    {X.columns[i]:<8}"
                      f"{r.importances_mean[i]:.3f}"
                      f" +/- {r.importances_std[i]:.3f}")
                
                

r2
    goals_z 0.891 +/- 0.089
    xg_z    0.056 +/- 0.017
    fouled_z0.029 +/- 0.011
    progruns_z0.029 +/- 0.010
    crosses_z0.012 +/- 0.006
neg_mean_absolute_percentage_error
    goals_z 0.659 +/- 0.066
    fouled_z0.039 +/- 0.016
    progruns_z0.028 +/- 0.012
    fouls_z 0.023 +/- 0.011
    tackles_z0.017 +/- 0.006
neg_mean_squared_error
    goals_z 66.643 +/- 6.680
    xg_z    4.222 +/- 1.289
    fouled_z2.158 +/- 0.859
    progruns_z2.150 +/- 0.757
    crosses_z0.932 +/- 0.433


- df.shape = (158, 17)
- Ridge Regression = 0.42629810615686115
- r2
    - goals_z 1.282 +/- 0.149
    - progruns_z0.096 +/- 0.029
    - xg_z    0.095 +/- 0.030
    - takeons_z0.026 +/- 0.013
- neg_mean_absolute_percentage_error
    - goals_z 0.988 +/- 0.115
- neg_mean_squared_error
    - goals_z 108.981 +/- 12.672
    - progruns_z8.139 +/- 2.494
    - xg_z    8.047 +/- 2.509
    - takeons_z2.195 +/- 1.067


- df.shape = (125, 17)

- Ridge Regression = 0.4494483686544971

- r2
    - goals_z 0.982 +/- 0.123
    - progruns_z0.132 +/- 0.039
    - xg_z    0.056 +/- 0.017
    - boxtouches_z0.046 +/- 0.023
    - nsxg_z  0.039 +/- 0.013
    - aerials_z0.031 +/- 0.014
- neg_mean_absolute_percentage_error
    - goals_z 0.888 +/- 0.122
    - progruns_z0.181 +/- 0.059
- neg_mean_squared_error
    - goals_z 89.125 +/- 11.206
    - progruns_z11.964 +/- 3.505
    - xg_z    5.055 +/- 1.543
    - boxtouches_z4.180 +/- 2.062
    - nsxg_z  3.509 +/- 1.182
    - aerials_z2.782 +/- 1.244

- df.shape = (145, 17)
- Ridge Regression = 0.43985733944194594

- r2
    - goals_z 0.919 +/- 0.122
    - progruns_z0.077 +/- 0.028
    - nsxg_z  0.069 +/- 0.026
    - boxtouches_z0.034 +/- 0.015
- neg_mean_absolute_percentage_error
    - goals_z 0.817 +/- 0.115
    - progruns_z0.124 +/- 0.036
    - tackles_z0.060 +/- 0.022
    - boxtouches_z0.048 +/- 0.023
- neg_mean_squared_error
    - goals_z 78.467 +/- 10.375
    - progruns_z6.549 +/- 2.364
    - nsxg_z  5.854 +/- 2.186
    - boxtouches_z2.943 +/- 1.292

- NMAPE = the higher the better
- NMSE
- R2 =  is a statistical measure of how close the data are to the fitted regression line . It is the percentage of variation (from 0 to 1) explained by the relationship between two variables.. metric that is used to evaluate the performance of a regression-based machine learning model.